_episodes/31-cmd-line.md

```diff
@@ -0,0 +1,532 @@
+
+~~~
+$ git init
+$ git add gdp_plots.sh
+$ git commit -m "First commit of analysis script"
+~~~
+{: .bash}
+
+Because we're only concerned with changes to our analysis script, we are going to
+create a .gitignore file for all of the gapminder `.csv` files and any Python notebook
+files (`.ipynb`) files we have created thus far.
+
+~~~
+$ echo "*.csv" > .gitignore
+$ echo "*.ipynb" >> .gitignore
+$ git add .gitignore
+$ git commit -m "Adding ignore file"
+~~~
+
+Now that we have a clean repository, let's get back to work on adding command line
+arguments to our program.
+
+## Changing code under Version Control
+
+As it is, this plot isn't bad but let's add some labels for clarity. We'll use the
+data filename as a title for the plot and indicate what information in on each axis.
+
+~~~
+import pandas
+# we need to import part of matplotlib
+# because we are no longer in a notebook
+import matplotlib.pyplot as plt
+
+filename = 'gapminder_gdp_oceania.csv'
+
+# read data into a pandas dataframe and transpose
+data = pandas.read_csv(filename, index_col = 'country').T
+
+# create a plot the transposed data
+ax = data.plot( title = filename )
+
+# set some plot attributes
+ax.set_xlabel("Year")
+ax.set_ylabel("GDP Per Capita")
+# set the x locations and labels
+ax.set_xticks( range(len(data.index)) )
+ax.set_xticklabels( data.index, rotation = 45 )
+
+# display the plot
+plt.show()
+~~~
+{: .python}
+
+Now when we run this, our plot looks a little bit nicer.
+
+~~~
+$ python gdp_plots.py
+~~~
+{: .bash}
+
+### Updating the Repository
+
+~~~
+$ git add gdp_plots.py
+$ git commit -m "Improving plot format"
+~~~
+{: .bash}
+
+## Command-Line Arguments
+
+This program currently only works for the Oceania set of data. How might we modify
+the program to work for any of the gapminder gdp data sets? We could go into the
+script and change the `.csv` filename to generate the same plot for different
+sets of data, but there is an even better way.
+
+Python programs can use additional arguments provided in the following manner.
+
+~~~
+$ python <program> <argument1> <argument2> <other_arguments>
+~~~
+{: .bash}
+
+The program can then use these arguments to alter its behavior based on those arguments.
+In this case, we'll be using arguments to tell our program to operate on a specific file.
+
+We'll be using the `sys` module to do so. `sys` (short for system) is a standard
+Python module used to store information about the program and its running
+environment, including what arguments were passed to the program when the
+command was executed. These arguments are stored as a list in `sys.argv`.
+
+These arguments can be accessed in our program by importing the `sys`
+module. The first argument in `sys.argv` is always the name of the program, so
+we'll find any additional arguments right after that in the list.
+
+Let's try this out in a separate location on your machine. The directory above
+our current directory (a.k.a. the home directory) should do nicely
```

Comment:
may not be true...

Reply:
Hrmmm I thought about having them just add this test program to the repo, but I'm also trying not to spend too much time on git. Would that be better?

Reply:
the idea is fine -- their home directory may just be one folder above.  But if the git repo is in the `data` folder, then just the level above should be fine.